### PR TITLE
Convert to Keras 3: `super_resolution_sub_pixel` example

### PR DESCRIPTION
**Pull request by awsaf49**
This PR will convert `Image Super-Resolution using an Efficient Sub-Pixel CNN` example to backend agnostic Keras 3. This example now runs smoothly on TF, Torch, JAX.

Note, original example uses `tf.nn.depth_to_space` but Keras 3 doesn't have any impl. of that function yet. Therefore, I implemented it manually with Keras 3 in a cell block.
```py
class DepthToSpace(layers.Layer):
    def __init__(self, block_size):
        super(DepthToSpace, self).__init__()
        self.block_size = block_size

    def call(self, input):
        batch, height, width, depth = ops.shape(input) 
        depth = depth // (self.block_size ** 2)

        x = ops.reshape(input, [batch, height, width, self.block_size, self.block_size, depth])
        x = ops.transpose(x, [0, 1, 3, 2, 4, 5])
        x = ops.reshape(x, [batch, height * self.block_size, width * self.block_size, depth])
        return x
```

### CODE REVIEW

**Comment by awsaf49** [comment]
@fchollet would love to hear your thoughts
Issue: #18468

Diff:
```diff
--- a/examples/keras_io/vision/super_resolution_sub_pixel.py
+++ b/examples/keras_io/vision/super_resolution_sub_pixel.py
@@ -1,21 +1,23 @@
 """
 Title: Image Super-Resolution using an Efficient Sub-Pixel CNN
 Author: [Xingyu Long](https://github.com/xingyu-long)
+Converted to Keras 3 by: [Md Awsfalur Rahman](https://awsaf49.github.io)
 Date created: 2020/07/28
 Last modified: 2020/08/27
 Description: Implementing Super-Resolution using Efficient sub-pixel model on BSDS500.
-Accelerator: GPU
 """
-
 """
 ## Introduction
 
 ESPCN (Efficient Sub-Pixel CNN), proposed by [Shi, 2016](https://arxiv.org/abs/1609.05158)
-is a model that reconstructs a high-resolution version of an image given a low-resolution version.
+is a model that reconstructs a high-resolution version of an image given a low-resolution
+version.
 It leverages efficient "sub-pixel convolution" layers, which learns an array of
 image upscaling filters.
 
-In this code example, we will implement the model from the paper and train it on a small dataset,
+In this code example, we will implement the model from the paper and train it on a small
+dataset,
+[BSDS500](https://www2.eecs.berkeley.edu/Research/Projects/CS/vision/grouping/resources.html).
 [BSDS500](https://www2.eecs.berkeley.edu/Research/Projects/CS/vision/grouping/resources.html).
 """
 
@@ -23,19 +25,21 @@
 ## Setup
 """
 
+import os
+os.environ["KERAS_BACKEND"] = "jax"
+
+import keras as keras
+from keras import layers
+from keras import ops
+from keras.utils import load_img
+from keras.utils import array_to_img
+from keras.utils import img_to_array
+from keras.preprocessing import image_dataset_from_directory
 import tensorflow as tf
 
-import os
 import math
 import numpy as np
 
-from tensorflow import keras
-from tensorflow.keras import layers
-from tensorflow.keras.utils import load_img
-from tensorflow.keras.utils import array_to_img
-from tensorflow.keras.utils import img_to_array
-from tensorflow.keras.preprocessing import image_dataset_from_directory
-
 from IPython.display import display
 
 """
@@ -46,7 +50,8 @@
 We use the built-in `keras.utils.get_file` utility to retrieve the dataset.
 """
 
-dataset_url = "http://www.eecs.berkeley.edu/Research/Projects/CS/vision/grouping/BSR/BSR_bsds500.tgz"
+dataset_url =
+"http://www.eecs.berkeley.edu/Research/Projects/CS/vision/grouping/BSR/BSR_bsds500.tgz"
 data_dir = keras.utils.get_file(origin=dataset_url, fname="BSR", untar=True)
 root_dir = os.path.join(data_dir, "BSDS500/data")
 
@@ -135,7 +140,6 @@ def scaling(input_image):
 and retrieve the `y` channel.
 """
 
-
 # Use TF Ops to process.
 def process_input(input, input_size, upscale_factor):
     input = tf.image.rgb_to_yuv(input)
@@ -180,10 +184,29 @@ def process_target(input):
 """
 
 
+class DepthToSpace(layers.Layer):
+    def __init__(self, block_size):
+        super(DepthToSpace, self).__init__()
+        self.block_size = block_size
+
+    def call(self, input):
+        batch, height, width, depth = ops.shape(input)
+        depth = depth // (self.block_size**2)
+
+        x = ops.reshape(
+            input, [batch, height, width, self.block_size, self.block_size, depth]
+        )
+        x = ops.transpose(x, [0, 1, 3, 2, 4, 5])
+        x = ops.reshape(
+            x, [batch, height * self.block_size, width * self.block_size, depth]
+        )
+        return x
+
+
 def get_model(upscale_factor=3, channels=1):
     conv_args = {
         "activation": "relu",
-        "kernel_initializer": "Orthogonal",
+        "kernel_initializer": "orthogonal",
         "padding": "same",
     }
     inputs = keras.Input(shape=(None, None, channels))
@@ -191,7 +214,7 @@ def get_model(upscale_factor=3, channels=1):
     x = layers.Conv2D(64, 3, **conv_args)(x)
     x = layers.Conv2D(32, 3, **conv_args)(x)
     x = layers.Conv2D(channels * (upscale_factor**2), 3, **conv_args)(x)
-    outputs = tf.nn.depth_to_space(x, upscale_factor)
+    outputs = DepthToSpace(upscale_factor)(x)
 
     return keras.Model(inputs, outputs)
 
@@ -313,11 +336,11 @@ def on_test_batch_end(self, batch, logs=None):
 
 early_stopping_callback = keras.callbacks.EarlyStopping(monitor="loss", patience=10)
 
-checkpoint_filepath = "/tmp/checkpoint"
+checkpoint_filepath = "/tmp/checkpoint.keras"
 
 model_checkpoint_callback = keras.callbacks.ModelCheckpoint(
     filepath=checkpoint_filepath,
-    save_weights_only=True,
+    save_weights_only=False,
     monitor="loss",
     mode="min",
     save_best_only=True,

```

File: examples/keras_io/vision/super_resolution_sub_pixel.py [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+
+Compared to the paper, we add one more layer and we use the `relu` activation function
+instead of `tanh`.
+It achieves better performance even though we train the model for fewer epochs.
+"""
+
+
+class DepthToSpace(layers.Layer):
+    def __init__(self, block_size):
+        super(DepthToSpace, self).__init__()
```
Comment (fchollet) [inline examples/keras_io/vision/super_resolution_sub_pixel.py:190]: Use the Python 3 style, `super().__init__()`

File: examples/keras_io/vision/super_resolution_sub_pixel.py [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+[BSDS500](https://www2.eecs.berkeley.edu/Research/Projects/CS/vision/grouping/resources.html).
+[BSDS500](https://www2.eecs.berkeley.edu/Research/Projects/CS/vision/grouping/resources.html).
+"""
+
+"""
+## Setup
+"""
+
+import os
+os.environ["KERAS_BACKEND"] = "jax"
```
Comment (fchollet) [inline examples/keras_io/vision/super_resolution_sub_pixel.py:30]: No need to hardcode a backend if it works with all backends -- you can remove these 2 lines

File: examples/keras_io/vision/super_resolution_sub_pixel.py [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+
+import keras as keras
+from keras import layers
+from keras import ops
+from keras.utils import load_img
+from keras.utils import array_to_img
+from keras.utils import img_to_array
+from keras.preprocessing import image_dataset_from_directory
+import tensorflow as tf
```
Comment (fchollet) [inline examples/keras_io/vision/super_resolution_sub_pixel.py:39]: Add a comment that this is only used for data preprocessing

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

**Comment by awsaf49** [comment]
@fchollet, Do you think `ops.nn.depth_to_space` (similar to `tf.nn.depth_to_space`) will be helpful in `keras`? In other words, should I create a PR on this?

File: examples/keras_io/vision/super_resolution_sub_pixel.py (hunk 3 above)
Comment (awsaf49) [reply examples/keras_io/vision/super_resolution_sub_pixel.py]: Added the comment.

File: examples/keras_io/vision/super_resolution_sub_pixel.py (hunk 2 above)
Comment (awsaf49) [reply examples/keras_io/vision/super_resolution_sub_pixel.py]: Okay.

File: examples/keras_io/vision/super_resolution_sub_pixel.py (hunk 1 above)
Comment (awsaf49) [reply examples/keras_io/vision/super_resolution_sub_pixel.py]: updated it.

**Comment by fchollet** [comment]
> Do you think ops.nn.depth_to_space (similar to tf.nn.depth_to_space) 

I'd say no, it's implementable as a simple combination of existing ops and it's a fairly niche use case.

**Review by fchollet — APPROVED** [review]
LGTM, thanks!

**Pull request merged**